synapse/util/caches/stream_change_cache.py

```diff
@@ -94,7 +95,17 @@ def get_entities_changed(
         """
         changed_entities = self.get_all_entities_changed(stream_pos)
         if changed_entities is not None:
-            result = set(changed_entities).intersection(entities)
+            # We now do an intersection, trying to do so in the most efficient
+            # way possible (some of these sets are *large*). First check in the
+            # given iterable is already set that we can reuse, otherwise we
+            # create a set of the *smallest* of the two iterables and call
+            # `intersection(..)` on it (this can be twice as fast as the reverse).
```

Comment:
https://wiki.python.org/moin/TimeComplexity#set has some stuff to say about this:
> Average case: O(min(len(s), len(t)) 
> Worst case: O(len(s) * len(t))
> 
> replace "min" with "max" if t is not a set

Reply:
Well that is interesting, I did:

```
In [1]: from synapse.util.stringutils import random_string                                                                                                                                                         

In [2]:                                                                                                                                                                                                            

In [2]: a = [random_string(10) for _ in range(10000)]                                                                                                                                                              

In [3]: b = [random_string(10) for _ in range(10)]                                                                                                                                                                 

In [4]: %timeit set(a).intersection(b)                                                                                                                                                                             
299 µs ± 3.41 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [5]: %timeit set(b).intersection(a)                                                                                                                                                                             
141 µs ± 9.19 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

I'm slightly surprised that they're symmetric when `t` is not a set

(I'm going to wait for you to point out my embarrassing typo now....)

Reply:
This looks sane to me. The only thing I can think of is that this is partially bound by the creation of the `set()` and not by the `intersction` code.

Reply:
Ah, that'd make sense if they ignored the time requirements for creating the set `s`

Reply:
Right, so it'd be interesting to see:

```pycon
>>> sa = set(a)
>>> sb = set(b)
>>> %timeit sa.interesection(b)
>>> %timeit sb.interesection(a)
```

Reply:
I looked at this a bit more and I think your conclusion is sane, but the gist of my exploration is that converting things to sets is expensive. 👍